File: Mage/ObservationEditCoordinator.m
//
//  ObservationEditCoordinator.m
//  MAGE
//
//  Created by Dan Barela on 8/10/17.
//  Copyright © 2017 National Geospatial Intelligence Agency. All rights reserved.
//

#import "ObservationEditCoordinator.h"
#import "Event.h"
#import "User.h"
#import "Attachment.h"
#import "ObservationPropertiesEditCoordinator.h"
#import "FormDefaults.h"
#import "MAGE-Swift.h"

@interface ObservationEditCoordinator() <ObservationPropertiesEditDelegate>

@property (strong, nonatomic) UIViewController *rootViewController;
@property (nonatomic, strong) NSManagedObjectContext *managedObjectContext;
@property (strong, nonatomic) Event *event;
@property (nonatomic) BOOL newObservation;
@property (strong, nonatomic) id<ObservationEditDelegate> delegate;
@property (strong, nonatomic) FormPickerViewController *formController;
@property (strong, nonatomic) UINavigationController *navigationController;
@property (strong, nonatomic) Observation *observation;
@property (strong, nonatomic) SFGeometry *location;
@property (strong, nonatomic) NSMutableArray *viewControllers;
@property (strong, nonatomic) NSDictionary *currentEditField;
@property (strong, nonatomic) id currentEditValue;
@property (strong, nonatomic) NSMutableArray *childCoordinators;
@property (nonatomic) CLLocationAccuracy accuracy;
@property (strong, nonatomic) NSString *provider;
@property (nonatomic) double delta;
@property (strong, nonatomic) ObservationPropertiesEditCoordinator *propertiesEditCoordinator;

@end

@implementation ObservationEditCoordinator

- (instancetype) initWithRootViewController: (UIViewController *) rootViewController andDelegate: (id<ObservationEditDelegate>) delegate andLocation: (SFGeometry *) location andAccuracy: (CLLocationAccuracy) accuracy andProvider: (NSString *) provider andDelta: (double) delta {
    self = [super init];
    if (!self) return nil;
    
    self.managedObjectContext = [NSManagedObjectContext MR_newMainQueueContext];
    self.managedObjectContext.parentContext = [NSManagedObjectContext MR_rootSavingContext];
    [self.managedObjectContext MR_setWorkingName:@"Observation New Context"];
    self.managedObjectContext.stalenessInterval = 0.0;
    
    self.newObservation = YES;
    self.observation = [self createObservationAtLocation:location withAccuracy:accuracy andProvider:provider andDelta:delta];
    
    [self setupCoordinatorWithRootViewController:rootViewController andDelegate:delegate];
    
    return self;
}

- (instancetype) initWithRootViewController:(UIViewController *)rootViewController andDelegate:(id<ObservationEditDelegate>)delegate andObservation:(Observation *)observation {
    self = [super init];
    if (!self) return nil;
    
    self.managedObjectContext = [NSManagedObjectContext MR_newMainQueueContext];
    self.managedObjectContext.parentContext = [NSManagedObjectContext MR_rootSavingContext];
    [self.managedObjectContext MR_setWorkingName:@"Observation Edit Context"];
    self.managedObjectContext.stalenessInterval = 0.0;
    
    self.observation = [self setupObservation: observation];
    [self setupCoordinatorWithRootViewController:rootViewController andDelegate:delegate];
    
    return self;
}

- (void) setupCoordinatorWithRootViewController: (UIViewController *) rootViewController andDelegate: (id<ObservationEditDelegate>) delegate {
    self.rootViewController = rootViewController;
    [self pushViewController:rootViewController];
    self.delegate = delegate;
    self.location = [self.observation getGeometry];
    self.childCoordinators = [[NSMutableArray alloc] init];
    self.navigationController = [[UINavigationController alloc] init];
}

- (Observation *) createObservationAtLocation: (SFGeometry *) location withAccuracy: (CLLocationAccuracy) accuracy andProvider: (NSString *) provider andDelta: (double) delta {
    self.newObservation = YES;
    Observation *observation = [Observation observationWithGeometry:location andAccuracy: accuracy andProvider: provider andDelta: delta inManagedObjectContext:self.managedObjectContext];
    observation.dirty = [NSNumber numberWithBool:YES];
    return observation;
}

- (Observation *) setupObservation: (Observation *) observation {
    Observation *observationInContext = [observation MR_inContext:self.managedObjectContext];
    observationInContext.dirty = [NSNumber numberWithBool:YES];
    return observationInContext;
}

- (NSMutableArray *) viewControllers {
    if (_viewControllers != nil) return _viewControllers;
    _viewControllers = [[NSMutableArray alloc] init];
    return _viewControllers;
}

- (UIViewController *) currentViewController {
    return [self.viewControllers lastObject];
}

- (void) popViewControllers {
    [self.viewControllers removeLastObject];
}

- (void) pushViewController: (UIViewController *) viewController {
    [self.viewControllers addObject:viewController];
}

- (Event *) event {
    if (_event) return _event;
    _event = [Event getCurrentEventInContext:self.managedObjectContext];
    return _event;
}

- (void) start {
    if (![self.event isUserInEvent:[User fetchCurrentUserInManagedObjectContext:self.managedObjectContext]]) {
        UIAlertController * alert = [UIAlertController
                                     alertControllerWithTitle:@"You are not part of this event"
                                     message:@"You cannot create observations for an event you are not part of."
                                     preferredStyle:UIAlertControllerStyleAlert];
        
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
        [self.rootViewController presentViewController:alert animated:YES completion:nil];
    } else {
        [self.navigationController setModalPresentationStyle:UIModalPresentationCustom];
        [self.navigationController setModalTransitionStyle:UIModalTransitionStyleCrossDissolve];
        [self.rootViewController presentViewController:self.navigationController animated:YES completion:^{

        }];
        [self startEditObservationFields];
        if (self.newObservation) {
            if ([self.event.nonArchivedForms count] > 1) {
                [self startFormPicker];
            } else if ([self.event.nonArchivedForms count] == 1) {
                [self addFormToObservation:[self.event.nonArchivedForms objectAtIndex:0]];
            }
        }
    }
}

- (void) addForm {
    [self startFormPicker];
}

- (void) startFormPicker {
    self.formController = [[FormPickerViewController alloc] initWithDelegate:self andForms:self.event.nonArchivedForms];
    
    UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:self.formController];
    UIBarButtonItem *backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Cancel" style:UIBarButtonItemStylePlain target:self action:@selector(cancelSelection)];
    [[self.formController navigationItem] setLeftBarButtonItem:backBarButtonItem];
    [[self navigationController] presentViewController:nav animated:true completion:nil];
}

- (void) startEditObservationFields {
    self.propertiesEditCoordinator = [[ObservationPropertiesEditCoordinator alloc] initWithObservation: self.observation andNewObservation: self.newObservation andNavigationController: self.navigationController andDelegate:self];
    [_childCoordinators addObject:self.propertiesEditCoordinator];
    [self.propertiesEditCoordinator start];
}

- (void) addFormToObservation: (NSDictionary *) form {
    NSMutableDictionary *newProperties = [self.observation.properties mutableCopy];
    
    NSMutableArray *observationForms = [[newProperties objectForKey:@"forms"] mutableCopy];
    NSMutableDictionary *newForm = [[NSMutableDictionary alloc] initWithObjectsAndKeys:[form objectForKey:@"id"], @"formId", nil];
    
    FormDefaults *defaults = [[FormDefaults alloc] initWithEventId:[self.observation.eventId integerValue] formId:[[form objectForKey:@"id"] integerValue]];
    NSDictionary *formDefaults = [defaults getDefaultsMap];
    
    NSArray *fields = [form objectForKey:@"fields"];
    if (formDefaults.count > 0) { // user defaults
        for (NSDictionary *field in fields) {
            id value = nil;
            id defaultField = [formDefaults objectForKey:[field objectForKey:@"id"]];
            if (defaultField) {
                value = [defaultField objectForKey:@"value"];
            }
            
            if (value) {
                [newForm setObject:value forKey:[field objectForKey:@"name"]];
            }
        }
    } else { // server defaults
        for (NSDictionary *field in fields) {
            // grab the server default from the form fields value property
            id value = [field objectForKey:@"value"];
            
            if (value) {
                [newForm setObject:value forKey:[field objectForKey:@"name"]];
            }
        }
    }
    
    [observationForms addObject:newForm];
    
    [newProperties setObject:observationForms forKey:@"forms"];
    self.observation.properties = newProperties;
    [self.propertiesEditCoordinator formAdded];
}

#pragma mark - FormPickedDelegate methods
- (void) formPicked:(NSDictionary *)form {
    [self.navigationController dismissViewControllerAnimated:YES completion:nil];
    NSLog(@"Form Picked %@", [form objectForKey:@"name"]);
    [self addFormToObservation:form];
//    [self startEditObservationFields];
}

- (void) cancelSelection {
    [self.navigationController dismissViewControllerAnimated:YES completion:^{
        NSLog(@"root view dismissed");
    }];
}

#pragma

#pragma mark - ObservationPropertiesEditDelegate methods

- (void) deleteObservation {
    [self.observation deleteObservationWithCompletion:^(BOOL contextDidSave, NSError * _Nullable error) {
        NSLog(@"Deleted");
    }];
    [self propertiesEditCanceled];
    [self.delegate observationDeleted:self.observation coordinator:self];
}

- (void) propertiesEditCanceled {
    self.managedObjectContext = nil;
    [self.delegate editCancel:self];
    [self.navigationController dismissViewControllerAnimated:YES completion:nil];
}

- (void) propertiesEditComplete {
    __weak typeof(self) weakSelf = self;
    
    self.observation.user = [User fetchCurrentUserInManagedObjectContext:self.managedObjectContext];
    
    [self.managedObjectContext MR_saveToPersistentStoreWithCompletion:^(BOOL contextDidSave, NSError *error) {
        if (!contextDidSave) {
            NSLog(@"Error saving observation to persistent store, context did not save");
        }
        
        if (error) {
            NSLog(@"Error saving observation to persistent store %@", error);
        }
        
        NSLog(@"saved the observation: %@", weakSelf.observation.remoteId);
        
        [weakSelf.delegate editComplete:weakSelf.observation coordinator:self];
        
        [_rootViewController dismissViewControllerAnimated:YES completion:^{
            NSLog(@"root view dismissed");
            weakSelf.managedObjectContext = nil;
        }];
    }];
}

#pragma


@end

File: Mage/ObservationEditCoordinator.swift
//
//  ObservationEditCoordinator.swift
//  MAGE
//
//  Created by Daniel Barela on 12/1/20.
//  Copyright © 2020 National Geospatial Intelligence Agency. All rights reserved.
//

import Foundation
import MaterialComponents.MaterialBottomSheet

@objc protocol ObservationEditListener {
    @objc func fieldValueChanged(_ field: [String: Any], value: Any?);
//    @objc optional func fieldSelected(_ field: [String: Any], currentValue: Any?);
    @objc optional func formUpdated(_ form: [String: Any], eventForm: [String: Any], form index: Int);
}

@objc protocol ObservationEditDelegate {
    @objc func editCancel(_ coordinator: NSObject);
    @objc func editComplete(_ observation: Observation, coordinator: NSObject);
}

protocol ObservationCommonPropertiesListener: AnyObject {
    func geometryUpdated(_ geometry: SFGeometry?, accuracy: String?, delta: Double?, provider: String?);
    func timestampUpdated(_ date: Date?);
}

@objc class ObservationEditCoordinator: NSObject {
    
    var newObservation = false;
    var observation: Observation?;
    weak var rootViewController: UIViewController?;
    var navigationController: UINavigationController?;
    weak var delegate: ObservationEditDelegate?;
    var observationEditController: ObservationEditCardCollectionViewController?;
    var observationFormReorder: ObservationFormReorder?;
    var bottomSheet: MDCBottomSheetController?;
    var currentEditField: [String: Any]?;
    var currentEditValue: Any?;
    var scheme: MDCContainerScheming?;
    
    private lazy var managedObjectContext: NSManagedObjectContext = {
        var managedObjectContext: NSManagedObjectContext = .mr_newMainQueue();
        managedObjectContext.parent = .mr_rootSaving();
        managedObjectContext.stalenessInterval = 0.0;
        managedObjectContext.mr_setWorkingName(newObservation ? "Observation New Context" : "Observation Edit Context");
        return managedObjectContext;
    } ()
    
    private lazy var event: Event? = {
        return Event.getCurrentEvent(context: self.managedObjectContext);
    } ()
    
    private lazy var eventForms: [[String: AnyHashable]] = {
        let eventForms = event?.forms as? [[String: AnyHashable]] ?? [];
        return eventForms;
    }()
    
    private lazy var user: User = {
        return User.fetchCurrentUser(in: self.managedObjectContext);
    }()
    
    @objc public func applyTheme(withContainerScheme containerScheme: MDCContainerScheming?) {
        self.scheme = containerScheme;
    }
    
    @objc public init(rootViewController: UIViewController!, delegate: ObservationEditDelegate, location: SFGeometry?, accuracy: CLLocationAccuracy, provider: String, delta: Double) {
        super.init();
        observation = createObservation(location: location, accuracy: accuracy, provider: provider, delta: delta);
        setupCoordinator(rootViewController: rootViewController, delegate: delegate);
    }
    
    @objc public init(rootViewController: UIViewController!, delegate: ObservationEditDelegate, observation: Observation) {
        super.init();
        self.observation = setupObservation(observation: observation);
        setupCoordinator(rootViewController: rootViewController, delegate: delegate);
    }
    
    @objc public func start() {
        guard let event = event else {
            return
        }

        if (!event.isUserInEvent(user: user)) {
            let alert = UIAlertController(title: "You are not part of this event", message: "You cannot create or edit observations for an event you are not part of.", preferredStyle: .alert);
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil));
            self.rootViewController?.present(alert, animated: true, completion: nil);
        } else {
            if let navigationController = navigationController {
                navigationController.modalPresentationStyle = .custom;
                navigationController.modalTransitionStyle = .crossDissolve;
                self.rootViewController?.present(navigationController, animated: true, completion: nil);
                observationEditController = ObservationEditCardCollectionViewController(delegate: self, observation: observation!, newObservation: newObservation, containerScheme: self.scheme);
                navigationController.pushViewController(observationEditController!, animated: true);
                if let scheme = self.scheme {
                    observationEditController?.applyTheme(withContainerScheme: scheme);
                }
            }
        }
    }
    
    @objc public func startFormReorder() {
        guard let event = event else {
            return
        }

        if (!event.isUserInEvent(user: user)) {
            let alert = UIAlertController(title: "You are not part of this event", message: "You cannot edit this observation.", preferredStyle: .alert);
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil));
            self.rootViewController?.present(alert, animated: true, completion: nil);
        } else {
            if let navigationController = navigationController {
                navigationController.modalPresentationStyle = .custom;
                navigationController.modalTransitionStyle = .crossDissolve;
                self.rootViewController?.present(navigationController, animated: true, completion: nil);
                observationFormReorder = ObservationFormReorder(observation: observation!, delegate: self, containerScheme: self.scheme);
                navigationController.pushViewController(self.observationFormReorder!, animated: true);
                self.observationFormReorder!.applyTheme(withContainerScheme: scheme);
            }
        }
    }
    
    func setupCoordinator(rootViewController: UIViewController, delegate: ObservationEditDelegate) {
        self.rootViewController = rootViewController;
        self.delegate = delegate;
        self.navigationController = UINavigationController();
    }
    
    func createObservation(location: SFGeometry?, accuracy: CLLocationAccuracy, provider: String, delta: Double) -> Observation {
        newObservation = true;
        let observation = Observation(geometry: location, andAccuracy: accuracy, andProvider: provider, andDelta: delta, in: managedObjectContext);
        observation.dirty = 1;
        addRequiredForms(observation: observation);
        return observation;
    }
    
    func setupObservation(observation: Observation) -> Observation? {
        let observationInContext = observation.mr_(in: self.managedObjectContext);
        observationInContext?.dirty = 1;
        return observationInContext;
    }
    
    func setupFormWithDefaults(observation: Observation, form: [String: Any]) -> [String: AnyHashable] {
        var newForm: [String: AnyHashable] = [EventKey.formId.key: form[FieldKey.id.key] as! Int];
        let defaults: FormDefaults = FormDefaults(eventId: observation.eventId as! Int, formId: form[FieldKey.id.key] as! Int);
        let formDefaults: [String: AnyHashable] = defaults.getDefaults() as! [String: AnyHashable];
        
        let fields: [[String : AnyHashable]] = (form[FormKey.fields.key] as! [[String : AnyHashable]]).filter { (($0[FieldKey.archived.key] as? Bool) == nil || ($0[FieldKey.archived.key] as? Bool) == false) };
        if (formDefaults.count > 0) { // user defaults
            for (_, field) in fields.enumerated() {
                var value: AnyHashable? = nil;
                if let defaultField: AnyHashable = formDefaults[field[FieldKey.name.key] as! String] {
                    value = defaultField
                }
                
                if (value != nil) {
                    newForm[field[FieldKey.name.key] as! String] = value;
                }
            }
        } else { // server defaults
            for (_, field) in fields.enumerated() {
                // grab the server default from the form fields value property
                if let value: AnyHashable = field[FieldKey.value.key] {
                    newForm[field[FieldKey.name.key] as! String] = value;
                }
            }
        }
        return newForm;
    }
    
    func addFormToObservation(observation: Observation, form: [String: AnyHashable]) {
        var observationProperties: [String: Any] = [ObservationKey.forms.key:[]];
        var observationForms: [[String: Any]] = [];
        if let properties = observation.properties as? [String: Any] {
            if (properties.keys.contains(ObservationKey.forms.key)) {
                observationForms = properties[ObservationKey.forms.key] as! [[String: Any]];
            }
            observationProperties = properties;
        }
        observationForms.append(setupFormWithDefaults(observation: observation, form: form));
        observationProperties[ObservationKey.forms.key] = observationForms;
        observation.properties = observationProperties;
    }
    
    func addRequiredForms(observation: Observation) {
        for eventForm in eventForms {
            let eventFormMin: Int = (eventForm[FieldKey.min.key] as? Int) ?? 0;
            if (eventFormMin > 0) {
                for _ in 1...eventFormMin {
                    addFormToObservation(observation: observation, form: eventForm);
                }
            }
        }
    }
}

extension ObservationEditCoordinator: FormPickedDelegate {
    func formPicked(form: [String : Any]) {
        observationEditController?.formAdded(form: form);
        bottomSheet?.dismiss(animated: true, completion: nil);
        bottomSheet = nil;
    }
    
    func cancelSelection() {
        bottomSheet?.dismiss(animated: true, completion: nil);
        bottomSheet = nil;
    }
}

extension ObservationEditCoordinator: FieldSelectionDelegate {
    func launchFieldSelectionViewController(viewController: UIViewController) {
        self.navigationController?.pushViewController(viewController, animated: true);
    }
}

extension ObservationEditCoordinator: ObservationFormReorderDelegate {
    func formsReordered(observation: Observation) {
        self.observation = observation;
        self.observation!.userId = user.remoteId;

        if let observationEditController = self.observationEditController {
            observationEditController.formsReordered(observation: self.observation!);
            self.navigationController?.popViewController(animated: true);

        } else {
            self.managedObjectContext.mr_saveToPersistentStore { [self] (contextDidSave, error) in
                if (!contextDidSave) {
                    print("Error saving observation to persistent store, context did not save");
                }
                
                if let safeError = error {
                    print("Error saving observation to persistent store \(safeError)");
                }
                
                delegate?.editComplete(self.observation!, coordinator: self as NSObject);

                self.navigationController?.dismiss(animated: true, completion: nil);
            }
        }
    }
    
    func cancelReorder() {
        if (self.observationEditController != nil) {
            self.navigationController?.popViewController(animated: true);
        } else {
            self.managedObjectContext.reset();
            self.navigationController?.dismiss(animated: true, completion: nil);
        }
    }
}

extension ObservationEditCoordinator: ObservationEditCardDelegate {    
    func reorderForms(observation: Observation) {
        self.observationFormReorder = ObservationFormReorder(observation: observation, delegate: self, containerScheme: self.scheme);
        self.navigationController?.pushViewController(self.observationFormReorder!, animated: true);
    }
    
    func addForm() {
        let forms: [[String: AnyHashable]] = (event?.forms as! [[String : AnyHashable]]).filter { form in
            return !(form[FormKey.archived.rawValue] as? Bool ?? false)
        };
        let formPicker: FormPickerViewController = FormPickerViewController(delegate: self, forms: forms, observation: observation, scheme: self.scheme);
        formPicker.applyTheme(withScheme: scheme);
        bottomSheet = MDCBottomSheetController(contentViewController: formPicker);
        bottomSheet?.trackingScrollView = formPicker.tableView
        self.navigationController?.present(bottomSheet!, animated: true, completion: nil);
    }
    
    func saveObservation(observation: Observation) {
        print("Save observation");
        self.observation!.userId = user.remoteId;
        self.managedObjectContext.mr_saveToPersistentStore { [self] (contextDidSave, error) in
            if (!contextDidSave) {
                print("Error saving observation to persistent store, context did not save");
            }
            
            if let safeError = error {
                print("Error saving observation to persistent store \(safeError)");
            }
            
            print("Saved the observation \(observation.remoteId ?? "")");
            delegate?.editComplete(observation, coordinator: self as NSObject);
            rootViewController?.dismiss(animated: true, completion: nil);
            observationEditController = nil;
            navigationController = nil;
        }
    }
    
    func cancelEdit() {
        print("Cancel the edit")
        let alert = UIAlertController(title: "Discard Changes", message: "Do you want to discard your changes?", preferredStyle: .alert);
        alert.addAction(UIAlertAction(title: "Yes, Discard", style: .destructive, handler: { [self] (action) in
            self.navigationController?.dismiss(animated: true, completion: nil);
            self.managedObjectContext.reset();
            delegate?.editCancel(self as NSObject);
            observationEditController = nil;
            navigationController = nil;
        }));
        alert.addAction(UIAlertAction(title: "No", style: .cancel, handler: nil));
        self.navigationController?.present(alert, animated: true, completion: nil);
    }
}
